sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+                  }
+                }
+
+                (Some(BucketSpec(numBuckets, bucketCols, sortCols, sortDirections)),
+                  false, false)
+              case Token("TOK_NOT_CLUSTERED", Nil) =>
+                (None, true, false)
+              case Token("TOK_NOT_SORTED", Nil) =>
+                (None, false, true)
+          }
+      }
+
+      AlterTableStoreProperties(
+        tableIdent,
+        buckets,
+        noClustered,
+        noSorted)(node.source)
+
+    case Token("TOK_ALTERTABLE_BUCKETS", Token(bucketNum, Nil) :: Nil) :: rest =>
+      val num = bucketNum.toInt
+      val buckets = Some(BucketSpec(num, Nil, Nil, Nil))
+      AlterTableStoreProperties(
+        tableIdent,
+        buckets,
+        false,
+        false)(node.source)
+
+    case (tableSkewed @ Token("TOK_ALTERTABLE_SKEWED", _)) :: rest =>
+      // Alter Table not skewed
+      // Token("TOK_ALTERTABLE_SKEWED", Nil) means not skewed.
+      val notSkewed = if (tableSkewed.children.size == 0) {
+        true
+      } else {
+        false
+      }
+
+      val (notStoredAsDirs, skewedArgs) = tableSkewed match {
+        case Token("TOK_ALTERTABLE_SKEWED", Token("TOK_STOREDASDIRS", Nil) :: Nil) =>
+          // Alter Table not stored as directories
+          (true, None)
+        case Token("TOK_ALTERTABLE_SKEWED", skewedArgs :: Nil) =>
+          val (cols, values, storedAsDirs) = skewedArgs match {
+            case Token("TOK_TABLESKEWED", skewedCols :: skewedValues :: stored) =>
+              val cols = skewedCols.children.map(n => unquoteString(cleanIdentifier(n.text)))
+              val values = skewedValues match {
+                case Token("TOK_TABCOLVALUE", values) =>
+                  Seq(values.map(n => unquoteString(cleanIdentifier(n.text))))
+                case Token("TOK_TABCOLVALUE_PAIR", pairs) =>
+                  pairs.map {
+                    case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                      values match {
+                        case Token("TOK_TABCOLVALUE", vals) =>
+                          vals.map(n => unquoteString(cleanIdentifier(n.text)))
+                      }
+                  }
+              }
+
+              val storedAsDirs = stored match {
+                case Token("TOK_STOREDASDIRS", Nil) :: Nil => true
+                case _ => false
+              }
+
+              (cols, values, storedAsDirs)
+          }
+          (false, Some((cols, values, storedAsDirs)))
+      }
+
+      if (skewedArgs.isDefined) {
+        AlterTableSkewed(
+          tableIdent,
+          skewedArgs.get._1, /* cols */
+          skewedArgs.get._2, /* values */
+          skewedArgs.get._3, /* storedAsDirs */
+          notSkewed, notStoredAsDirs)(node.source)
+      } else {
+        AlterTableSkewed(tableIdent, Nil, Nil, false, notSkewed, notStoredAsDirs)(node.source)
+      }
+
+    case Token("TOK_ALTERTABLE_SKEWED_LOCATION", args) :: rest =>
+      val skewedMaps = args(0) match {
```

Comment:
Can there be more than one location? You could do a direct pattern match, i.e:

    val Seq(Token("TOK_SKEWED_LOCATIONS", Token("TOK_SKEWED_LOCATION_LIST", locationMaps) :: Nil)) = args